### PR TITLE
Fix PVF precompilation for Kusama

### PR DESCRIPTION
**Pull request by AndreiEres**
image

Because on Kusama validators.len() < discovery_keys.len() we can tweak the PVF precompilation to allow prepare PVFs when the node is an authority but not a validator.

### CODE REVIEW

**Review by eskimor — APPROVED** [review]
Function name could be more descriptive, other than that: Looks good.

File: polkadot/node/core/candidate-validation/src/lib.rs [hunk 1]
```diff
@@ -427,14 +427,15 @@ where
...
+	let is_present_validator = session_info
+		.validators
 		.iter()
-		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]));
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), ValidatorId::ID)]));
 
 	// There is still a chance to be a previous session authority, but this extra work does not
 	// affect the finalization.
-	is_past_present_or_future_authority && !is_present_authority
+	is_past_present_or_future_authority && !is_present_validator
```
Comment (sandreim) [inline polkadot/node/core/candidate-validation/src/lib.rs:438]: This distinction exists until we remove the `max_validators` configuration so even without this fix it still works as expected on Polkadot.

**Review by sandreim — APPROVED** [review]

File: polkadot/node/core/candidate-validation/src/lib.rs (hunk 1 above)
Comment (AndreiEres) [reply polkadot/node/core/candidate-validation/src/lib.rs:438]: Sure

**Review by ordian — APPROVED** [review]
Looking at what `check_next_session_authority`, I understand this change, but it seems to conflate multiple cases in a not-ideal way.

My understanding is that after restart we want to check if we are a (para)validator in the current or the next session, and if so, precompile PVFs.
Also, if we are already running and are not a (para)validator in the current session, but are in the next, we also want to precompile PVFs.

What this function does is checks for (not-only-para)validator set of current and next session along with a paravalidator set of past `dispute_period` sessions. It doesn't distinguish fresh start from already running and checks if we are not current session (with this change para)validator. 
Ideally, we'd have a separate runtime API to tell us if we are in the next (para)validator set, but we don't have that currently, so I understand this is a workaround solution. We could also probably simplify the check to just look at the next set assuming that if we already have the artifacts cached, pre-compilation will be a no-op.

**Comment by ordian** [comment]
Missing prdoc

**Review by alexggh — APPROVED** [review]

**Pull request merged**